File: api/app/interfaces/schemas/codebase.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, model_validator

from app.domain.execution.run import RunStatus
from app.domain.models.codebase import (
    ArtifactFormat,
    ArtifactKind,
    CodebaseSourceType,
    CodebaseStatus,
    FileTreeNode,
    SessionMode,
    SymbolKind,
)
from app.domain.models.codebase_version import CodebaseVersionState


class CreateCodebaseRequest(BaseModel):
    name: str = "未命名代码库"
    source_type: CodebaseSourceType = CodebaseSourceType.FILES
    file_id: str | None = None
    git_url: str | None = None
    file_ids: list[str] | None = Field(default_factory=list)

    @model_validator(mode="after")
    def validate_source_payload(self) -> "CreateCodebaseRequest":
        if self.source_type is CodebaseSourceType.ZIP and not self.file_id:
            raise ValueError("zip source requires file_id")
        if self.source_type is CodebaseSourceType.GIT and not self.git_url:
            raise ValueError("git source requires git_url")
        if self.source_type is CodebaseSourceType.FILES and not self.file_ids:
            raise ValueError("files source requires at least one file_id")
        return self


class CodebaseResponse(BaseModel):
    id: str
    name: str
    source_type: CodebaseSourceType
    source_ref: str = ""
    status: CodebaseStatus
    language_stats: dict[str, int] = Field(default_factory=dict)
    file_count: int = 0
    sandbox_id: str | None = None
    workspace_path: str = ""
    error: str | None = None
    vector_degraded: bool = False
    active_version_id: str | None = None
    created_at: datetime
    updated_at: datetime


class ListCodebasesResponse(BaseModel):
    codebases: list[CodebaseResponse]


class FileTreeResponse(BaseModel):
    tree: list[FileTreeNode]


class SymbolResponse(BaseModel):
    id: str
    name: str
    kind: SymbolKind
    file_id: str
    path: str = ""
    signature: str = ""
    start_line: int = 0
    end_line: int = 0
    parent_id: str | None = None


class ListSymbolsResponse(BaseModel):
    symbols: list[SymbolResponse]


class ArtifactResponse(BaseModel):
    id: str
    version_id: str | None = None
    kind: ArtifactKind
    format: ArtifactFormat
    title: str
    content: str
    meta: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class ListArtifactsResponse(BaseModel):
    artifacts: list[ArtifactResponse]


class ReadSourceRequest(BaseModel):
    path: str
    start_line: int | None = None
    end_line: int | None = None


class ReadSourceResponse(BaseModel):
    path: str
    content: str
    start_line: int | None = None
    end_line: int | None = None


class CodebaseBuildResponse(BaseModel):
    id: str
    run_id: str | None = None
    codebase_id: str
    version_id: str
    status: RunStatus
    phase: str | None = None
    progress: int = 0
    failure_code: str | None = None
    created_at: datetime
    updated_at: datetime
    terminal_at: datetime | None = None
    can_retry: bool = False
    can_cancel: bool = False


class CodebaseVersionResponse(BaseModel):
    id: str
    codebase_id: str
    parent_version_id: str | None = None
    build_id: str
    state: CodebaseVersionState
    source_snapshot_key: str | None = None
    source_revision: str = ""
    source_digest: str = ""
    capabilities: dict[str, Any] = Field(default_factory=dict)
    degraded_reasons: list[str] = Field(default_factory=list)
    metrics: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    published_at: datetime | None = None
    is_active: bool = False
    is_published: bool = False
    is_candidate: bool = False
    build: CodebaseBuildResponse | None = None


class ListCodebaseVersionsResponse(BaseModel):
    codebase_id: str
    active_version_id: str | None = None
    active_build: CodebaseBuildResponse | None = None
    versions: list[CodebaseVersionResponse] = Field(default_factory=list)


class CreateCodebaseSessionRequest(BaseModel):
    mode: SessionMode = SessionMode.ASK
    model_id: str | None = None
    skill_id: str | None = None
    codebase_version_id: str | None = None


class CreateCodebaseSessionResponse(BaseModel):
    session_id: str
    codebase_id: str
    mode: SessionMode


class DownloadCodebaseResponse(BaseModel):
    snapshot_key: str
    download_url: str = ""

File: ui/src/lib/api/types/codebase.ts
import type { SessionMode } from "./session";

// ==================== 代码知识库 ====================

export type CodebaseStatus =
  | "pending"
  | "materializing"
  | "analyzing"
  | "indexing"
  | "generating"
  | "ready"
  | "failed";

export type CodebaseSourceType = "zip" | "git" | "files";

type ArtifactKind =
  | "architecture"
  | "data_flow"
  | "module_dir"
  | "flowchart"
  | "call_chain"
  | "overview";

export type Codebase = {
  id: string;
  name: string;
  source_type: CodebaseSourceType;
  source_ref?: string;
  status: CodebaseStatus;
  language_stats?: Record<string, number>;
  file_count?: number;
  sandbox_id?: string | null;
  workspace_path?: string;
  ingest_task_id?: string | null;
  error?: string | null;
  vector_degraded?: boolean;
  active_version_id?: string | null;
  created_at?: string;
  updated_at?: string;
};

export type CodebasesData = {
  codebases: Codebase[];
};

export type FileTreeNode = {
  name: string;
  path: string;
  is_dir: boolean;
  language?: string;
  children?: FileTreeNode[];
};

export type FileTreeData = {
  tree: FileTreeNode[];
};

export type CodebaseSymbol = {
  id: string;
  name: string;
  kind: string;
  file_id: string;
  path?: string;
  signature?: string;
  start_line?: number;
  end_line?: number;
  parent_id?: string | null;
};

export type CodebaseSymbolsData = {
  symbols: CodebaseSymbol[];
};

export type CodebaseArtifact = {
  id: string;
  version_id?: string | null;
  kind: ArtifactKind;
  format: "mermaid" | "markdown";
  title: string;
  content: string;
  meta?: Record<string, unknown>;
  created_at?: string;
};

export type CodebaseArtifactsData = {
  artifacts: CodebaseArtifact[];
};

export type CreateCodebaseParams = {
  name?: string;
  source_type: CodebaseSourceType;
  file_id?: string;
  git_url?: string;
  file_ids?: string[];
};

export type CreateCodebaseSessionParams = {
  mode?: SessionMode;
  model_id?: string;
  skill_id?: string;
  codebase_version_id?: string;
};

export type CodebaseSessionData = {
  session_id: string;
  codebase_id: string;
  mode: SessionMode;
};

export type ReadSourceParams = {
  path: string;
  start_line?: number;
  end_line?: number;
};

export type ReadSourceData = {
  path: string;
  content: string;
  start_line?: number;
  end_line?: number;
};

export type DownloadCodebaseData = {
  snapshot_key: string;
  download_url?: string;
};

type CodebaseBuildState =
  | "queued"
  | "running"
  | "succeeded"
  | "degraded"
  | "failed"
  | "cancelled";

type CodebaseVersionState = "building" | "ready" | "degraded" | "failed";

export type CodebaseBuild = {
  id: string;
  codebase_id: string;
  version_id: string;
  parent_version_id?: string | null;
  command_key: string;
  state: CodebaseBuildState;
  phase?: string | null;
  progress: number;
  capabilities: unknown[];
  degraded_reasons: unknown[];
  metrics: Record<string, unknown>;
  error_code?: string | null;
  error_message?: string | null;
  heartbeat_at?: string | null;
  last_event_seq: number;
  created_at: string;
  started_at?: string | null;
  finished_at?: string | null;
  can_retry: boolean;
  can_cancel: boolean;
};

export type CodebaseVersion = {
  id: string;
  codebase_id: string;
  parent_version_id?: string | null;
  build_id?: string | null;
  state: CodebaseVersionState;
  source_snapshot_key?: string | null;
  source_revision?: string;
  source_digest?: string;
  capabilities: Record<string, unknown>;
  degraded_reasons: string[];
  metrics: Record<string, unknown>;
  created_at: string;
  published_at?: string | null;
  is_active: boolean;
  is_published: boolean;
  is_candidate: boolean;
  build?: CodebaseBuild | null;
};

export type CodebaseVersionsData = {
  codebase_id: string;
  active_version_id?: string | null;
  active_build?: CodebaseBuild | null;
  versions: CodebaseVersion[];
};
